### PR TITLE
Statistical assertions for debugging

### LINKED ISSUE #2866: Statistical assertions for debugging
### What is the expected enhancement?
Quantum programs are hard to debug - unlike classical programs, we can't print quantum states.  We measure intermediate states by creating a copy of the quantum circuit up to that point in the code, then use statistical tests to verify that the distribution of measurement results is as expected.
Pull request incoming that implements ideas in [this paper](http://yipenghuang.com/wp-content/uploads/2019/06/p541-huang.pdf)

### PR DESCRIPTION
**Pull request by the-legend-of-lia**
Fixes #2866 

### Summary
Issue #2866: Statistical assertions for debugging
* This pull request contains 74 commits, 4 new files in the codebase in a new folder qiskit/assertions, and adds several methods to qiskit/result/result.py to get assertion results.
* This implements the ideas in this paper: http://yipenghuang.com/wp-content/uploads/2019/06/p541-huang.pdf

### Details and comments
* All code, examples in examples/python/assertions, and test cases in test/python/assertions are documented and linted.
* All test cases in test/python/assertions pass.
* Additional examples are in forks of qiskit-aqua and qiskit-tutorials-community, to be PR'd after this PR.

### CODE REVIEW

**Comment by the-legend-of-lia** [comment]
#2866

**Comment by edasgupta** [comment]
Hi! One of the emails I committed with is a school email that doesn't actually have an inbox so I haven't been able to sign the CLA with it yet. I am working with my university to fix this.

File: examples/python/assertions/hello_quantum.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2018.
```
Comment (1ucian0) [inline examples/python/assertions/hello_quantum.py]:
```suggestion
# (C) Copyright IBM 2019.
```

File: examples/python/assertions/qft.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017.
```
Comment (1ucian0) [inline examples/python/assertions/qft.py]:
```suggestion
# (C) Copyright IBM 2019.
```

File: examples/python/assertions/teleport.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017.
```
Comment (1ucian0) [inline examples/python/assertions/teleport.py]:
```suggestion
# (C) Copyright IBM 2019.
```

File: examples/python/assertions/using_qiskit_terra_level_0_qc1.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2018.
```
Comment (1ucian0) [inline examples/python/assertions/using_qiskit_terra_level_0_qc1.py]:
```suggestion
# (C) Copyright IBM 2019.
```

File: examples/python/assertions/using_qiskit_terra_level_0_qc2.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2018.
```
Comment (1ucian0) [inline examples/python/assertions/using_qiskit_terra_level_0_qc2.py]:
```suggestion
# (C) Copyright IBM 2019.
```

File: examples/python/assertions/using_qiskit_terra_level_1.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2018.
```
Comment (1ucian0) [inline examples/python/assertions/using_qiskit_terra_level_1.py]:
```suggestion
# (C) Copyright IBM 2019.
```

File: qiskit/assertions/__init__.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017.
```
Comment (1ucian0) [inline qiskit/assertions/__init__.py]:
```suggestion
# (C) Copyright IBM 2019.
```

File: qiskit/assertions/assertclassical.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017.
```
Comment (1ucian0) [inline qiskit/assertions/assertclassical.py]:
```suggestion
# (C) Copyright IBM 2019.
```

File: qiskit/assertions/assertproduct.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017.
```
Comment (1ucian0) [inline qiskit/assertions/assertproduct.py]:
```suggestion
# (C) Copyright IBM 2019.
```

File: qiskit/assertions/asserts.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017.
```
Comment (1ucian0) [inline qiskit/assertions/asserts.py]:
```suggestion
# (C) Copyright IBM 2019.
```

File: qiskit/assertions/assertuniform.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017.
```
Comment (1ucian0) [inline qiskit/assertions/assertuniform.py]:
```suggestion
# (C) Copyright IBM 2019.
```

File: test/python/assertions/__init__.py [hunk 12]
```diff
@@ -0,0 +1,15 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2018.
```
Comment (1ucian0) [inline test/python/assertions/__init__.py]:
```suggestion
# (C) Copyright IBM 2019.
```

File: test/python/assertions/test_assert_classical.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2018.
```
Comment (1ucian0) [inline test/python/assertions/test_assert_classical.py]:
```suggestion
# (C) Copyright IBM 2019.
```

File: test/python/assertions/test_assert_product.py [hunk 14]
```diff
@@ -0,0 +1,78 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2018.
```
Comment (1ucian0) [inline test/python/assertions/test_assert_product.py]:
```suggestion
# (C) Copyright IBM 2019.
```

File: test/python/assertions/test_assert_uniform.py [hunk 15]
```diff
@@ -0,0 +1,77 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2018.
```
Comment (1ucian0) [inline test/python/assertions/test_assert_uniform.py]:
```suggestion
# (C) Copyright IBM 2019.
```

File: qiskit/result/result.py [hunk 16]
```diff
@@ -181,6 +182,87 @@ def get_counts(self, experiment=None):
...
+        new_counts = {}
+        for (key, value) in counts.items():
+            new_key = key.replace(' ', '')
+            new_key = ''.join([new_key[-1*(cbit+1)] for cbit in cbits][::-1])
+            new_counts.setdefault(new_key, 0)
+            new_counts[new_key] += value
+        counts = new_counts
+
+        chisq, pval, passed = assertion.stat_test(counts)
+        return chisq, pval, passed if not assertion._negate else not passed
```
Comment (1ucian0) [inline qiskit/result/result.py]: Would it be better to return an `AssertionResult` instance instead of a tuple?

File: qiskit/result/result.py [hunk 17]
```diff
@@ -181,6 +182,87 @@ def get_counts(self, experiment=None):
...
+
+    def get_assertion_pval(self, experiment=None):
+        """
+        Calculate and collect results of statistical tests for an experiment.
+        Args:
+            experiment (QuantumCircuit): a breakpoint
+        Returns:
+            float: p-value of statistical test
+        """
+        _, pval, _ = self.get_assertion(experiment)
```
Comment (1ucian0) [inline qiskit/result/result.py]: This method seems unnecessary.

File: qiskit/result/result.py [hunk 18]
```diff
@@ -181,6 +182,87 @@ def get_counts(self, experiment=None):
...
+        """
+        Calculate and collect results of statistical tests for an experiment.
+
+        Args:
+            experiment (QuantumCircuit): a breakpoint
+
+        Returns:
+            float: chi-squared test statistic
+        """
+        chisq, _, _ = self.get_assertion(experiment)
```
Comment (1ucian0) [inline qiskit/result/result.py]: This method seems unnecessary.

File: qiskit/result/result.py [hunk 19]
```diff
@@ -181,6 +182,87 @@ def get_counts(self, experiment=None):
...
+        """
+        Calculate and collect results of statistical tests for an experiment.
+
+        Args:
+            experiment (QuantumCircuit): a breakpoint
+
+        Returns:
+            Boolean: if assertion passed
+        """
+        _, _, passed = self.get_assertion(experiment)
```
Comment (1ucian0) [inline qiskit/result/result.py]: This method seems unnecessary.

File: qiskit/assertions/assertproduct.py [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+            qubit1(QuantumRegister or list or tuple): quantum register
+            cbit1(ClassicalRegister or list or tuple): classical register
+            pcrit(float): the critical p-value
+            negate(bool): True if assertion passed is negation of statistical test passed
+        """
+        self._qubit0 = self.syntax4measure(qubit0)
+        self._cbit0 = self.syntax4measure(cbit0)
+        self._qubit1 = self.syntax4measure(qubit1)
+        self._cbit1 = self.syntax4measure(cbit1)
+        type_str = "Not Product" if negate else "Product"
```
Comment (1ucian0) [inline qiskit/assertions/assertproduct.py]: Maybe have classes for different assertion types?

File: qiskit/assertions/asserts.py [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+    @staticmethod
+    def breakpoint_name():
+        """
+        Returns the name of the breakpoint.  A breakpoint is a QuantumCircuit
+        whose last instruction is an assertion instruction.
+
+        Returns:
+            string: the breakpoint name
+        """
+        return "breakpoint_" + datetime.now().isoformat()
```
Comment (1ucian0) [inline qiskit/assertions/asserts.py]: The `breakpoint_name` changes dynamically?

File: qiskit/result/result.py [hunk 22]
```diff
@@ -181,6 +182,87 @@ def get_counts(self, experiment=None):
...
+        """
+        Return the type of the assertion for an experiment.
+
+        Args:
+            experiment (QuantumCircuit): a breakpoint
+        Returns:
+            string: the type of assertion
+        """
+        assertion = experiment.data[-1][0]
+        return assertion._type
```
Comment (1ucian0) [inline qiskit/result/result.py]: if assertions of different types are different classes, the `_type` is unnecessary.

File: qiskit/result/result.py [hunk 23]
```diff
@@ -181,6 +182,87 @@ def get_counts(self, experiment=None):
...
+            experiment (QuantumCircuit): a breakpoint
+
+        Returns:
+            tuple: tuple containing:
+
+                chisq(float): the chi-squared value
+                pval(float): the p-value
+                passed(Boolean): if assertion passed
+        """
+        exp = experiment
```
Comment (1ucian0) [inline qiskit/result/result.py]: it seems unnecessary

File: qiskit/result/result.py [hunk 24]
```diff
@@ -181,6 +182,87 @@ def get_counts(self, experiment=None):
...
+        cbits = assertion._cbit
+        cbits = Asserts.clbits2idxs(cbits, exp)
+
+        new_counts = {}
+        for (key, value) in counts.items():
+            new_key = key.replace(' ', '')
+            new_key = ''.join([new_key[-1*(cbit+1)] for cbit in cbits][::-1])
+            new_counts.setdefault(new_key, 0)
+            new_counts[new_key] += value
+        counts = new_counts
```
Comment (1ucian0) [inline qiskit/result/result.py]: it seems unnecessary

File: qiskit/assertions/asserts.py [hunk 25]
```diff
@@ -0,0 +1,120 @@
...
+        else:
+            return [bit]
+
+    @staticmethod
+    def clbits2idxs(cbits, exp):
+        """
+        Returns the indices of cbits with respect to exp, the QuantumCircuit.
+
+        Args:
+            cbits(ClassicalRegister or list or tuple): cbit of a measurement instruction
```
Comment (1ucian0) [inline qiskit/assertions/asserts.py]: bits as tuples is being deprecated.

File: qiskit/assertions/asserts.py [hunk 26]
```diff
@@ -0,0 +1,120 @@
...
+            counts(dictionary): result.get_counts(experiment)
+
+        Returns:
+            tuple: tuple containing:
+
+                chisq(float): the chi-squared value
+                pval(float): the p-value
+                passed(bool): if the test passed
+        """
+        return
```
Comment (1ucian0) [inline qiskit/assertions/asserts.py]: maybe `raise NotImplementedError`?

File: qiskit/assertions/assertproduct.py [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+        qubit0(QuantumRegister or list or tuple): quantum register
+        cbit0(ClassicalRegister or list or tuple): classical register
+        qubit1(QuantumRegister or list or tuple): quantum register
+        cbit1(ClassicalRegister or list or tuple): classical register
+        pcrit(float): critical p-value for the hypothesis test
+
+    Returns:
+        QuantumCircuit: copy of quantum circuit at the assert point
+    """
+    clone = self.copy(Asserts.breakpoint_name())
```
Comment (1ucian0) [inline qiskit/assertions/assertproduct.py]: if `self` is a QuantumCircuit, why not change it in-place? `get_breakpoint_not_product` could be called `append_breakpoint_not_product`.

File: qiskit/assertions/assertuniform.py [hunk 28]
```diff
@@ -0,0 +1,110 @@
...
+    Returns:
+        QuantumCircuit: copy of quantum circuit at the assert point
+    """
+    clone = self.copy(Asserts.breakpoint_name())
+    assertion = AssertUniform(qubit, cbit, pcrit, False)
+    clone.append(assertion, [assertion._qubit], [assertion._cbit])
+    return clone
+
+
+QuantumCircuit.get_breakpoint_uniform = get_breakpoint_uniform
```
[the remark this answers is not in the document]
Comment (the-legend-of-lia) [reply qiskit/assertions/assertuniform.py]: This code style is consistent with the Measure class, which is the grandparent class of all assertion classes.

File: qiskit/assertions/assertuniform.py [hunk 29]
```diff
@@ -0,0 +1,110 @@
...
+    Returns:
+        QuantumCircuit: copy of quantum circuit at the assert point
+    """
+    clone = self.copy(Asserts.breakpoint_name())
+    assertion = AssertUniform(qubit, cbit, pcrit, True)
+    clone.append(assertion, [assertion._qubit], [assertion._cbit])
+    return clone
+
+
+QuantumCircuit.get_breakpoint_not_uniform = get_breakpoint_not_uniform
```
[the remark this answers is not in the document]
Comment (the-legend-of-lia) [reply qiskit/assertions/assertuniform.py]: This code style is consistent with the Measure class, which is the grandparent class of all assertion classes.

File: qiskit/result/result.py (hunk 16 above)
Comment (the-legend-of-lia) [reply qiskit/result/result.py]: @yipenghuang0302

**Comment by 1ucian0** [comment]
Some went wrong in a merge?

File: qiskit/assertions/assertproduct.py (hunk 20 above)
Comment (the-legend-of-lia) [reply qiskit/assertions/assertproduct.py]: There are different classes; each assertion type (AssertClassical, AssertProduct, and AssertUniform) and its logical negation are the same type except for the value of the boolean negate.
The type string is for the user to print out what type of assertion they had run, so given an assertion result (passed or failed), they can easily tell whether it was a "Not Product" assertion or a "Product" assertion.

File: qiskit/assertions/assertproduct.py (hunk 27 above)
Comment (the-legend-of-lia) [reply qiskit/assertions/assertproduct.py]: A breakpoint is a deepcopy of the QuantumCircuit, with an assertion instruction tacked onto the end.  The reason for copying the QuantumCircuit is to allow multiple breakpoints and the original QuantumCircuit to be executed.  Since the assertion is a measure instruction, appending (instead of copying and appending) would alter the behavior of the remainder of the QuantumCircuit.

File: qiskit/result/result.py (hunk 24 above)
Comment (the-legend-of-lia) [reply qiskit/result/result.py]: This enables assertions to be performed on parts of a result.
The counts dictionary of a result is for all cbits in the quantum circuit.
Say you wanted to make an assertion on just some of the qubits in the quantum circuit, and not all of them, this piece of code builds a new counts dictionary just for the cbits the assertion is on.
Thus, for a state that is |0>|+>|1>, you can assertclassical that the 1st and 3rd qubits are "01", and assertsuperposition on the 2nd qubit.

**Comment by the-legend-of-lia** [comment]
The reason for the force push was because I can't sign the CLA until my email used for recent commits is set up, so I git filter-branch to retroactively change my commit email address.  However, the force push modified commit history for 1000+ commits not in this fork, so I undid it.

File: qiskit/assertions/asserts.py (hunk 21 above)
Comment (the-legend-of-lia) [reply qiskit/assertions/asserts.py]: [resolved by this commit](https://github.com/Qiskit/qiskit-terra/pull/2867/commits/1a040382dfc8495c056438c110bd4a68b7d85f2c)

File: qiskit/result/result.py (hunk 23 above)
Comment (the-legend-of-lia) [reply qiskit/result/result.py]: [resolved by this commit](https://github.com/Qiskit/qiskit-terra/pull/2867/commits/1a040382dfc8495c056438c110bd4a68b7d85f2c)

File: qiskit/assertions/asserts.py (hunk 25 above)
Comment (the-legend-of-lia) [reply qiskit/assertions/asserts.py]: [resolved by this commit](https://github.com/Qiskit/qiskit-terra/pull/2867/commits/e64d7ea155c224da82759685a0c44e59aa11758a)

File: qiskit/assertions/asserts.py (hunk 26 above)
Comment (the-legend-of-lia) [reply qiskit/assertions/asserts.py]: [resolved by this commit](https://github.com/Qiskit/qiskit-terra/pull/2867/commits/e64d7ea155c224da82759685a0c44e59aa11758a)

File: qiskit/result/result.py (hunk 19 above)
Comment (the-legend-of-lia) [reply qiskit/result/result.py]: This is to allow this 1 line syntax:
assert( result.get_assertion_passed(breakpoint) )

Without this method, this would require 2 lines:
_, _, passed = result.get_assertion(breakpoint)
assert(passed)

File: qiskit/result/result.py (hunk 22 above)
Comment (the-legend-of-lia) [reply qiskit/result/result.py]: The type string is for the user to print out what type of assertion they had run, so given an assertion result (passed or failed), they can easily tell whether it was a "Not Product" assertion or a "Product" assertion.

File: qiskit/result/result.py (hunk 17 above)
Comment (the-legend-of-lia) [reply qiskit/result/result.py]: [resolved by this commit](https://github.com/edasgupta/qiskit-terra/commit/6cd3d4157a2fa31a4cd53eeb029c29df8b9cd37c)

File: qiskit/result/result.py (hunk 18 above)
Comment (the-legend-of-lia) [reply qiskit/result/result.py]: [resolved by this commit](https://github.com/edasgupta/qiskit-terra/commit/6cd3d4157a2fa31a4cd53eeb029c29df8b9cd37c)

File: qiskit/result/result.py (hunk 16 above)
Comment (the-legend-of-lia) [reply qiskit/result/result.py]: (https://github.com/edasgupta/qiskit-terra/commit/3bd8cb5725605232238dfffee1bf4f6cc972662b)
now get_assertion gets the assertion object, and this method was renamed to get_assertion_stats

File: qiskit/result/result.py (hunk 24 above)
Comment (1ucian0) [reply qiskit/result/result.py]: What I mean is `chisq, pval, passed = assertion.stat_test(new_counts)`....

Comment (1ucian0) [reply qiskit/result/result.py]: Also, consider to remove the `get_` in the methods names https://github.com/Qiskit/qiskit-terra/issues/1346

**Comment by the-legend-of-lia** [comment]
A note on the above failed tests: By their nature, statistical tests have a false positive and false negative probability that are anti-correlated.  The purpose of this debugging tool is to allow a programmer to catch bugs by consistently failing these statistical tests.  When a coding error is made, from what I've seen, it is not uncommon to see statistical tests having p-values less than 10^(-14).

The tests thus far have critical p-values of 0.01, which is 10^12 times larger than 10^(-10).  This means that the probability of seeing a false positive is very small (catching programming errors much quicker) at the cost of seeing way more false negatives.  This is why we recommend programmers use large critical p-values for unit-testing, and smaller critical p-values for integration testing.

tl;dr I made the mistake of sticking with unnecessarily large critical p-values for these tests and am going to push changes for that shortly, with an estimate of how often they'll fail.

More on that:
For a given p-value, the assert_uniform test is more prone to have false negatives (return False for the statistical test when the distribution was programmed to be uniform) than the other 2 types of assertions because superposition states have more quantum error than quantum states.

File: qiskit/result/result.py (hunk 24 above)
Comment (yipenghuang0302) [reply qiskit/result/result.py]: This is fixed by 954105393d451637bf3076c066dd86afccbe5000

**Comment by 1ucian0** [comment]
This PR is stalled for some time already.

Here is the situation. The PR adds a fully independent feature (it's a literal addition, no change, no remove). It could be fully implemented independently as a qiskit extension. After several months of this PR in the queue, it seems not being a priority to take over the ownership of this code, at least not for now.

I think this code should be maintained by the original author in their own repo.

I planning a dashboard to list and track qiskit-related code (such like this one) as a way to highlight them and improve their maintainability (for example, with early warnings when we deprecate code on qiskit). I would love if the authors of this PR want to participate as the first external module on this dashboard.

The discussion about the integration of this feature into qiskit terra should happen in #2866

Thanks!
